functional/kata-monitor/run.sh

```diff
@@ -0,0 +1,245 @@
+PODID=""
+CONID=""
+
+FALSE=1
+TRUE=0
+
+title() {
+	local STEP="$1"
+	echo "* STEP: $STEP"
+}
+
+echo_ok() {
+	local MSG="$1"
+
+	echo "OK: $MSG"
+	echo ""
+}
+
+error_with_msg() {
+	local MSG=${1}
+
+	echo -e "\nERROR: $MSG"
+	echo -e "\nkata-monitor logs:\n----------------"
+	cat "$MONITOR_LOG_FILE"
+	cleanup
+	echo -e "\nkata-monitor testing: FAILED!"
+	exit 1
+}
+
+cleanup() {
+	crictl stop "$CONID" "$VANILLA_CONID"  > /dev/null 2>&1
+	crictl rm "$CONID" "$VANILLA_CONID"    > /dev/null 2>&1
+	crictl stopp "$PODID" "$VANILLA_PODID" > /dev/null 2>&1
+	crictl rmp "$PODID" "$VANILLA_PODID"   > /dev/null 2>&1
+	kill -9 "$KATA_MONITOR_PID" > /dev/null 2>&1
+	rm -rf "$TMPATH"
+}
+
+create_sandbox_json() {
+	local sbfile="$1"
+
+	POD_NAME="$(date | md5sum | cut -d ' ' -f 1)"
+
+	/bin/cat <<OEM >$sbfile
+{
+	"metadata": {
+		"name": "nginx-$POD_NAME",
+		"namespace": "default",
+		"attempt": 1
+	},
+	"logDirectory": "/tmp",
+	"linux": {
+	}
+}
+OEM
+}
+
+create_container_json() {
+	local cntfile="$1"
+
+	/bin/cat <<OEM >$cntfile
+{
+	"metadata": {
+		"name": "busybox"
+	},
+	"image":{
+		"image": "busybox"
+	},
+	"command": [
+		"top"
+	],
+	"log_path":"busybox.log",
+	"linux": {
+	}
+}
+OEM
+}
+
+start_workload() {
+	local runtime=${1}
+	local args=""
+	local sbfile="${TMPATH}/sandbox.json"
+	local cntfile="${TMPATH}/container.json"
+
+	[ -n "$runtime" ] && args="-r $runtime"
+
+	create_sandbox_json "$sbfile"
+	create_container_json "$cntfile"
+
+	PODID=$(crictl runp $args $sbfile)
+	[ $? -ne 0 ] && error_with_msg "cannot create sandbox"
+	CONID=$(crictl create $PODID $cntfile $sbfile)
+	[ $? -ne 0 ] && error_with_msg "cannot create container"
+	crictl start $CONID > /dev/null
+	[ $? -ne 0 ] && error_with_msg "cannot start container $CONID"
+}
+
+stop_workload() {
+	crictl stop $CONID > /dev/null
+	[ $? -ne 0 ] && error_with_msg "cannot stop container $CONID"
+	crictl rm $CONID > /dev/null
+	[ $? -ne 0 ] && error_with_msg "cannot remove container $CONID"
+	crictl stopp $PODID > /dev/null
+	[ $? -ne 0 ] && error_with_msg "cannot stop pod $PODID"
+	crictl rmp $PODID > /dev/null
+	[ $? -ne 0 ] && error_with_msg "cannot remove pod $PODID"
+}
+
+is_sandbox_there() {
+	local podid=${1}
+	local sbs s
+
+	sbs=$(curl -s 127.0.0.1:8090/sandboxes)
+	[ $? -ne 0 ] && error_with_msg "cannot retrieve the pod list"
+	if [ -n "$sbs" ]; then
+		for s in $sbs; do
+			if [ "$s" = "$podid" ]; then
+				return $TRUE
+				break
+			fi
+		done
+	fi
+	return $FALSE
+}
+
+is_sandbox_there_iterate() {
+	local podid=${1}
+
+	for i in $(seq 1 $CACHE_UPD_TIMEOUT_SEC); do
+		is_sandbox_there "$podid" && return $TRUE
+		echo -n "."
+		sleep 1
+		continue
+	done
+
+	return $FALSE
+}
+
+is_sandbox_missing_iterate() {
+	local podid=${1}
+
+	for i in $(seq 1 $CACHE_UPD_TIMEOUT_SEC); do
+		is_sandbox_there "$podid" || return $TRUE
+		echo -n "."
+		sleep 1
+		continue
+	done
+
+	return $FALSE
+}
+
+###########################
+title "pre-checks"
```

Comment:
Since the rest of the script is nicely divided into functions, I'd consider adding a `main` for completeness and consistency:

```bash
main()
{
    # ...
}

main "$@"
```